### DISCUSSION TITLE
Code changes / Bug fixes to be included in sandbox  /glade/work/fvitt/camdev/carma_trop_stratv10

### DISCUSSION THREAD

**Discussion by tilmes** [NCAR/CARMA_dev#22] [trop_strat CARMA model]
Summary of bug fixes and location of the code based on current sandbox:
See https://github.com/NCAR/CARMA_dev/discussions/16 for details
**Code with the fixed for trop_strat_soa1**
/glade/p/cesm/chwg_dev/tilmes/cases/carma/ISAMIP_CESM2CARMA_WACCMMA2000climo.carma_trop_strat.feb10.all_fix_aqchem/SourceMods/src.cam and /glade/p/cesm/chwg_dev/tilmes/cases/carma/ISAMIP_CESM2CARMA_WACCMMA2000climo.carma_trop_strat.feb10.all_fix_aqchem/user_nl_cam
@fvitt Would you be able to add these changes and create a new sandbox version 10?

1. Dust emission fix, that caused extra sulfate emissions: Change in carma_model_mod.F90  -> problem only for  CARMA
<img width="635" alt="image" src="https://user-images.githubusercontent.com/22222754/218274993-0a173bbe-d410-428b-b7a4-aabf501f76de.png">
to 
<img width="684" alt="image" src="https://user-images.githubusercontent.com/22222754/218275002-81842e81-ab1a-4539-ad93-04c5e216f0fc.png">

2. Increase of cloud-borne sulfate counting production from interstitial to cloudborne 3times due to macmic time loop. Problem occurs for both CARMA and MAM. Temporary fix is the change of the number to substeps in mg and macmic loop through namelist changes:
cld_macmic_num_steps = 1
 micro_mg_num_steps = 3

3. Extra production of sulfate through advection  (mostly in the larger bins), only a CARMA problem due to the use of a concentration element in CARMA.
Fix: Added a global sulfate fixer and changed the original fixer into a column fixer that tries to conserve mass within the column.The global fixer calculates the total positive sulfate values and the total of the negative sulfate values, and then tries to fill in the negative values with positive values from elsewhere. To adjust the positive values, a global scaling factor was created that is (total_positive - abs(total_negative)) / (total_positive) for each bin.add a global sulfate fixer and changed the original fixer into a column fixer that tries to conserve mass within the column.The global fixer calculates the total positive sulfate values and the total of the negative sulfate values, and then tries to fill in the negative values with positive values from elsewhere. To adjust the positive values, I create a global scaling factor that is (total_positive - abs(total_negative)) / (total_positive) for each bin.
(changes in various parts)

4. Aqueous-phase chemistry fix to remove reactions on ice: Affects both MAM and CARMA: change in chemsitry.F90

5. Add budget and tracer diagnostics to identify changes in both interstitial and cloudborne tendencies for various parts in the code. This will be a useful diagnostic for future model developments.
Change in namelist setting:  carma_diags_file determines if and where extra output for diagnostics is added. Number 1 means it is added to fincl1, number 2: added to fincle2 etc.
 carma_do_package_diags = .true.
 carma_do_budget_diags =  .true.
 carma_diags_file = 2
 carma_diags_packages = "EFIX", "CRGFIX", "DRYADJ", "DEEP", "SHALLOW", "CLUBB01", "MICROP01",
    "WETDEPA", "CRFIXW", "WETDEPC", "DEEP2", "SURFACE", "CHEMEMIS",
    "CREMIS", "VDIF", "CHEM", "DRYDEPA", "CRTEND", "GWAVE", "NUDGE", "DMEADJ"

Summary of changed routines:
aero_check_mod.F90  aero_model.F90  carma_aero_convproc.F90  carma_constants_mod.F90  carma_fixer_mod.F90  carma_flags_mod.F90  carma_intr.F90  carma_model_mod.F90  chemistry.F90  physics_types.F90  physpkg.F90, namelist_definition.xml

More changes for trop_strat_soa5 will follow.

**Comment by tilmes** [comment]
Updated to the new sandbox with reference to the above:

2. This change turns out to be not needed any more, since it was not a problem, but just a diagnostics problem in CARMA
5. The tendency package has been updated and would be added this way in the namelist to be added to fincl2
carma_do_budget_diags = .true.
carma_do_package_diags = .true.
carma_diags_file = 2
carma_diags_packages = "CRGFIX", "MICROP01", "MICROP02", "MICROP03",
    "WETDEPA", "CREMIS", "VDIF", "CHEM", "CRTEND"
carma_debug_packages = "aero_model_wetdep"

Additional updates:
We also updated both MAM and CARMA code to include a long-wave absorption output for the 10micron wavelength band. The new namelist variable is called: 'AODABS10um'

Important variables to be added for budgets:
'H2SO4TC','DH2SO4CHM' (H2SO4 tendencies in CARMA, and net chemical production/loss of H2SO4 in CAM)

@cbardeen It would be useful to summarize the tendency budgets you added, including fixer changes. Would you mind adding this in this thread?

**Reply by cbardeen** [reply]
Not exactly sure what you want added, but maybe this will cover it.

As far as fixer changes, this refers to fixing situations where the sum of the core masses os greater than than the mass of the particle, which implies a negative mass for the concentration element that is physically impossible. The prior code had fixers that would replace negative values with 0. This had the effect of adding mass for the concentration element (sulfate), particularly from horizontal advection. To fix horizontal advection a global mass fixer has been added in carma_intr that is called after advection and will set the negative values to 0 by globally taking mass from where it is positive to balance the amount being added and thus conserve mass. In addition, a column mass fixer was added in carma_base that will move mass in the column to remove the negatives and conserve mass. There is still a gridbox fixer used in microfast that removes negative values, but is not mass conserving.

As to the diagnostic, new interfaces have been add to carma_intr and carma_model_mod to allow reporting of more diagnostics from physpkg and to allow the particular model to extend the diagnostics that can be reported. For the trop-strat model, diagnostics fields were added to calculate the SO4 burden both incloud and after instrumented physics packages. They are the the total burden (i.e. globally for the total column and including all bins). This is done instantaneously, so that burdens do not suffer from trying to determine them using monthly average values. There are namelist fields that control the behaviour of the diagnostics as well as the code that was added to carma_model_mod.

<entry id="carma_do_budget_diags" type="logical" category="carma"
       group="carma_nl" valid_values="" >
Flag indicating that model specific budget diagnostics should be
generated.
Default: FALSE
</entry>

<entry id="carma_do_package_diags" type="logical" category="carma"
       group="carma_nl" valid_values="" >
Flag indicating that model specific budget diagnostics should be
generated per physics package for the packages listed in
carma_diag_packages.
Default: FALSE
</entry>

<entry id="carma_diags_packages" type="char*12(100)" category="carma"
       group="carma_nl" valid_values="" >
List of physics packages for which diagnostic output is desired.
NOTE: These are shortname abbreviations for the paackages, so that
they can be used as part of history field names. See physpkg.f90 for
the names that are used.
Default: NONE
</entry>

<entry id="carma_debug_packages" type="char*32(100)" category="carma"
       group="carma_nl" valid_values="" >
List of physics packages for which debug output from the local carma
state checker is desired.
NOTE: This are the package names used in ptend, and can be long. See
physics_types.f90
Default: NONE
</entry>

<entry id="carma_diags_file" type="integer" category="carma"
       group="carma_nl" valid_values="" >
When > 0, indicates the history file to be used by default for
diagnostic output. A value of 1 indicated the h0 file.
Default: 0
</entry>

The budget diagnostics currently provided are:

     call addfld("SO4PRBD", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial SO4 pure burden')
      if (carma_diags_file > 0) call add_default("SO4PRBD", carma_diags_file, ' ')
      call addfld("SO4MXBD", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial SO4 mix burden')
      if (carma_diags_file > 0) call add_default("SO4MXBD", carma_diags_file, ' ')
      call addfld("SO4PRCLDBD", horiz_only, 'A', 'kg/m2', 'CARMA, Cloudborne SO4 pure burden')
      if (carma_diags_file > 0) call add_default("SO4PRCLDBD", carma_diags_file, ' ')
      call addfld("SO4MXCLDBD", horiz_only, 'A', 'kg/m2', 'CARMA, Cloudborne SO4 mix burden')

      if (carma_diags_file > 0) call add_default("SO4MXCLDBD", carma_diags_file, ' ')
      call addfld("SO4PRSF", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial SO4 pure surface flux')
      if (carma_diags_file > 0) call add_default("SO4PRSF", carma_diags_file, ' ')
      call addfld("SO4MXSF", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial SO4 mix surface flux')
      if (carma_diags_file > 0) call add_default("SO4MXSF", carma_diags_file, ' ')

      call addfld("H2SO4BD", horiz_only, 'A', 'kg/m2', 'CARMA, H2SO4 burden')
      if (carma_diags_file > 0) call add_default("H2SO4BD", carma_diags_file, ' ')
      call addfld("SO2BD", horiz_only, 'A', 'kg/m2', 'CARMA, SO2 burden')
      if (carma_diags_file > 0) call add_default("SO2BD", carma_diags_file, ' ')

      call addfld("MXBCBD", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial BC mix burden')
      if (carma_diags_file > 0) call add_default("MXBCBD", carma_diags_file, ' ')
      call addfld("MXDUSTBD", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial Dust mix burden')
      if (carma_diags_file > 0) call add_default("MXDUSTBD", carma_diags_file, ' ')
      call addfld("MXOCBD", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial OC mix burden')
      if (carma_diags_file > 0) call add_default("MXOCBD", carma_diags_file, ' ')
      call addfld("MXSALTBD", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial Sea Salt mix burden')
      if (carma_diags_file > 0) call add_default("MXSALTBD", carma_diags_file, ' ')
      call addfld("MXSOABD", horiz_only, 'A', 'kg/m2', 'CARMA, Interstitial SOA mix burden')
      if (carma_diags_file > 0) call add_default("MXSOABD", carma_diags_file, ' ')

      call addfld("MXBCCLDBD", horiz_only, 'A', 'kg/m2', 'CARMA, Cloudborne BC mix burden')
      if (carma_diags_file > 0) call add_default("MXBCCLDBD", carma_diags_file, ' ')
      call addfld("MXDUSTCLDBD", horiz_only, 'A', 'kg/m2', 'CARMA, Cloudborne Dust mix burden')
      if (carma_diags_file > 0) call add_default("MXDUSTCLDBD", carma_diags_file, ' ')
      call addfld("MXOCCLDBD", horiz_only, 'A', 'kg/m2', 'CARMA, Cloudborne OC mix burden')
      if (carma_diags_file > 0) call add_default("MXOCCLDBD", carma_diags_file, ' ')
      call addfld("MXSALTCLDBD", horiz_only, 'A', 'kg/m2', 'CARMA, Cloudborne Sea Salt mix burden')
      if (carma_diags_file > 0) call add_default("MXSALTCLDBD", carma_diags_file, ' ')
      call addfld("MXSOACLDBD", horiz_only, 'A', 'kg/m2', 'CARMA, Cloudborne SOA mix burden')
      if (carma_diags_file > 0) call add_default("MXSOACLDBD", carma_diags_file, ' ')


The package diagnostics are:

        call addfld("SO4PRBD_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2', trim(carma_diags_packages(i))//', SO4 pure burden')
        if (carma_diags_file > 0) call add_default("SO4PRBD_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')
        call addfld("SO4MXBD_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2', trim(carma_diags_packages(i))//', SO4 mixed burden')
        if (carma_diags_file > 0) call add_default("SO4MXBD_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')

        call addfld("SO4PRSF_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2/s', trim(carma_diags_packages(i))//', Surface Flux, SO4 pure tendency')
        if (carma_diags_file > 0) call add_default("SO4PRSF_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')
        call addfld("SO4MXSF_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2/s', trim(carma_diags_packages(i))//', Surface Flux, SO4 mix tendency')
        if (carma_diags_file > 0) call add_default("SO4MXSF_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')

        call addfld("SO4PRTC_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2/s', trim(carma_diags_packages(i))//', SO4 pure tendency')
        if (carma_diags_file > 0) call add_default("SO4PRTC_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')
        call addfld("SO4MXTC_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2/s', trim(carma_diags_packages(i))//', SO4 mixed tendency')
        if (carma_diags_file > 0) call add_default("SO4MXTC_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')

        call addfld("SO4PRCLDBD_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2', trim(carma_diags_packages(i))//', Cloudborne SO4 pure burden')
        if (carma_diags_file > 0) call add_default("SO4PRCLDBD_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')
        call addfld("SO4MXCLDBD_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2', trim(carma_diags_packages(i))//', Cloudborne SO4 mixed burden')
        if (carma_diags_file > 0) call add_default("SO4MXCLDBD_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')

        call addfld("SO4PRCLDTC_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2/s', trim(carma_diags_packages(i))//', Cloudborne SO4 pure tendency')
        if (carma_diags_file > 0) call add_default("SO4PRCLDTC_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')
        call addfld("SO4MXCLDTC_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2/s', trim(carma_diags_packages(i))//', Cloudborne SO4 mixed tendency')
        if (carma_diags_file > 0) call add_default("SO4MXCLDTC_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')

        call addfld("H2SO4TC_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2/s', trim(carma_diags_packages(i))//', H2SO4 total tendency')
        if (carma_diags_file > 0) call add_default("H2SO4TC_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')
        call addfld("SO2TC_"//trim(carma_diags_packages(i)), horiz_only, 'A', 'kg/m2/s', trim(carma_diags_packages(i))//', SO2 total tendency')
        if (carma_diags_file > 0) call add_default("SO2TC_"//trim(carma_diags_packages(i)), carma_diags_file, ' ')

If using the latest merge from Francis, I would recommend setting carma_debug_packages = "all", which should cause statistics to be kept for where mass fixing is done and how big the corrections are. This can provide a sense of the magnitude of the issue. It does not include the fixing done internal to CARMA, just the fixing after other physics packages. It also does not currently include the global mass fixer used after advection, which is probably the biggest source but is unavoidable. Values from other packages could represent errors in the code; although, they aren't necessarily errors.

If you use the settings listed by Simone, you should be able to use a jupyter phyton script of mine that esstimates the budget based on the package statistics. There is also on that uses the standard diagnostic output; although, it does require the output of a large number of fields to be able to do that calculation for both interstitial and cloudborne aerosol. You can look at the scripts to identify the fields they are using. You can get the scripts here:

/glade/work/bardeenc/notebooks/projects/carma/simone_budget_packages.ipynb
/glade/work/bardeenc/notebooks/projects/carma/simone_budget_diagnostic.ipynb

I also did one for a MAM run that used the diagnostic output that Simone had available. It was missing some terms like the chmical production of H2SO4.

/glade/work/bardeenc/notebooks/projects/carma/simone_budget_diagnostic-mam.ipynb

**Reply by cbardeen** [reply]
For the WACCM version (single SOA element), there was also an error where only the SOA element was getting updated and not the total mass. This cause a removal of sulfate equal to the amount of the SOA addition and could have created situations where the sum of the core masses was greater than the concentration element.

I also added some diagnostics from CARMA for the SOA growth and photolytic losses:

    ! Provide diagnostics on the SOA tendencies that affect MXAER.
    do ibin = 1, NBIN
      write(binname, '(A, I2.2)') "MXSOA", ibin     

      call addfld(trim(binname)//"CM", (/ 'lev' /), 'A', 'kg/kg/s', 'MXAER SOA gas condensation tendency')
      call addfld(trim(binname)//"PT", (/ 'lev' /), 'A', 'kg/kg/s', 'MXAER SOA photolysis tendency')
    end do

These are per bin and are not added automatically to the output.

**Comment by tilmes** [comment]
The new sandbox has now been created: /glade/work/fvitt/camdev/carma_trop_strat10.  (March 10th, 2023)

**Comment by tilmes** [comment]
Sandbox /glade/work/fvitt/camdev/carma_trop_stratv12 works with different compiler and new LW AOD output.
